google/cloud/spanner/client.cc

```diff
@@ -128,6 +128,37 @@ std::shared_ptr<Connection> MakeConnection(
                                                     std::move(stub));
 }
 
+StatusOr<CommitResult> RunTransaction(
+    Client client, Transaction::ReadWriteOptions const& opts,
+    std::function<TransactionAction(Client, Transaction)> const& f) {
+  Transaction txn = MakeReadWriteTransaction(opts);
+  TransactionAction action;
+#if GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
+  try {
+#endif
+    action = f(client, txn);
```

Comment:
As discussed briefly offline, I think the caller may be able to give us a std::function that accepts these args as *non-const* references. And if that's the case, the caller's functor *could* change, or even move-from the `client` and `txn` arg. This would be a problem if we needed to commit or rollback the transaction, or in a retry situation, reusing the `client`.

So, I think an easy fix would be to always pass the copies here like:

```cc
action = f(Client(client), Transaction(txn));
```

Reply:
As also briefly discussed offline, we're convinced that an argument `std::function` that accepts its args as non-const references would not be convertible to the `RunTransaction()` parameter.